File: entre_mc/entre_mc/page/painel_de_credito/painel_de_credito.js
// Copyright (c) 2026, Dércio Bobo and contributors
// For license information, please see license.txt

frappe.pages["painel-de-credito"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Painel de Crédito",
		single_column: true,
	});

	new EntreMcPainel(page);
};

const CURRENCY = "MZN";

class EntreMcPainel {
	constructor(page) {
		this.page = page;
		this.controls = {};
		this.render_shell();
		this.make_filtros();
		this.carregar();
	}

	render_shell() {
		this.inject_styles();

		this.page.main.html(`
			<div class="emc-painel">
				<section class="emc-painel-filtros" id="emc-painel-filtros">
					<button class="btn btn-default btn-sm emc-painel-limpar">${__("Limpar Filtros")}</button>
				</section>

				<section class="emc-resumo-cards" id="emc-painel-cards"></section>

				<div class="emc-painel-tabelas">
					<section>
						<h4>${__("Carteira em Risco")}</h4>
						<div class="emc-resumo-table-wrap">
							<table class="emc-resumo-table">
								<thead>
									<tr>
										<th>${__("Pedido")}</th>
										<th>${__("Cliente")}</th>
										<th class="emc-num">${__("Saldo do Crédito")}</th>
										<th class="emc-num">${__("Dívida")}</th>
										<th class="emc-num">${__("Em Risco")}</th>
									</tr>
								</thead>
								<tbody id="emc-painel-risco-rows"></tbody>
							</table>
						</div>
					</section>

					<section>
						<h4>${__("Créditos em Atraso")}</h4>
						<div class="emc-resumo-table-wrap">
							<table class="emc-resumo-table">
								<thead>
									<tr>
										<th>${__("Pedido")}</th>
										<th>${__("Cliente")}</th>
										<th>${__("Nº")}</th>
										<th>${__("Data Limite")}</th>
										<th class="emc-num">${__("Dias de Atraso")}</th>
										<th class="emc-num">${__("Total em Atraso")}</th>
									</tr>
								</thead>
								<tbody id="emc-painel-atraso-rows"></tbody>
							</table>
						</div>
					</section>
				</div>
			</div>
		`);

		this.$filtros = this.page.main.find("#emc-painel-filtros");
		this.$cards = this.page.main.find("#emc-painel-cards");
		this.$riscoRows = this.page.main.find("#emc-painel-risco-rows");
		this.$atrasoRows = this.page.main.find("#emc-painel-atraso-rows");

		this.$filtros.find(".emc-painel-limpar").on("click", () => {
			Object.values(this.controls).forEach((control) => control.set_value(""));
		});
	}

	make_filtros() {
		const campos = [
			{ fieldname: "cliente", label: __("Cliente"), fieldtype: "Link", options: "Cliente" },
			{ fieldname: "produto", label: __("Produto"), fieldtype: "Link", options: "Produto" },
			{
				fieldname: "finalidade",
				label: __("Finalidade"),
				fieldtype: "Select",
				options: "\nComercio\nAgricultura\nPecuaria\nIndústria\nServiços\nConsumo\nOutros",
			},
			{ fieldname: "data_inicio", label: __("Desembolsos de"), fieldtype: "Date" },
			{ fieldname: "data_fim", label: __("até"), fieldtype: "Date" },
		];

		campos.forEach((df) => {
			const $field = $(`<div class="emc-painel-filtro"></div>`).insertBefore(
				this.$filtros.find(".emc-painel-limpar")
			);
			const control = frappe.ui.form.make_control({
				df: { ...df, change: () => this.on_filtro_change() },
				parent: $field,
				render_input: true,
			});
			control.refresh();
			this.controls[df.fieldname] = control;
		});
	}

	on_filtro_change() {
		clearTimeout(this._timeout);
		this._timeout = setTimeout(() => this.carregar(), 300);
	}

	valores_filtro() {
		const valores = {};
		Object.entries(this.controls).forEach(([fieldname, control]) => {
			const value = control.get_value();
			if (value) valores[fieldname] = value;
		});
		return valores;
	}

	carregar() {
		frappe.call({
			method: "entre_mc.entre_mc.page.painel_de_credito.painel_de_credito.obter_painel",
			args: this.valores_filtro(),
			callback: (r) => r.message && this.render(r.message),
		});
	}

	render(data) {
		this.render_cards(data.cards);
		this.render_tabela_risco(data.carteira_em_risco);
		this.render_tabela_atraso(data.creditos_em_atraso);
	}

	render_cards(cards) {
		const items = [
			{ label: __("Saldo do Crédito"), value: format_currency(cards.saldo_do_credito, CURRENCY) },
			{
				label: __("Dívida"),
				value: format_currency(cards.divida, CURRENCY),
				warn: flt(cards.divida) > 0,
			},
			{
				label: __("Em Risco"),
				value: format_currency(cards.em_risco, CURRENCY),
				warn: flt(cards.em_risco) > 0,
			},
			{
				label: __("PAR %"),
				value: `${flt(cards.par_percent).toFixed(1)}%`,
				warn: flt(cards.par_percent) > 0,
			},
			{ label: __("Créditos Ativos"), value: cint(cards.num_ativos) },
			{
				label: __("Incumprimentos"),
				value: cint(cards.num_incumprimento),
				warn: cint(cards.num_incumprimento) > 0,
			},
			{ label: __("Desembolsado no Período"), value: format_currency(cards.desembolsado_periodo, CURRENCY) },
			{ label: __("Taxas Cobradas no Período"), value: format_currency(cards.taxas_periodo, CURRENCY) },
		];

		this.$cards.html(
			items
				.map(
					(c) => `
					<div class="emc-resumo-card${c.warn ? " emc-resumo-card--warn" : ""}">
						<span class="emc-resumo-card-label">${c.label}</span>
						<span class="emc-resumo-card-value">${c.value}</span>
					</div>`
				)
				.join("")
		);
	}

	render_tabela_risco(rows) {
		if (!rows || !rows.length) {
			this.$riscoRows.html(
				`<tr><td colspan="5" class="text-muted">${__("Sem créditos em risco para os filtros selecionados.")}</td></tr>`
			);
			return;
		}
		this.$riscoRows.html(
			rows
				.map(
					(r) => `
					<tr>
						<td><a href="/app/pedido-de-credito/${encodeURIComponent(r.name)}">${frappe.utils.escape_html(r.name)}</a></td>
						<td>${frappe.utils.escape_html(r.cliente || "")}</td>
						<td class="emc-num">${format_currency(r.saldo_do_credito, CURRENCY)}</td>
						<td class="emc-num">${format_currency(r.divida, CURRENCY)}</td>
						<td class="emc-num">${format_currency(r.em_risco, CURRENCY)}</td>
					</tr>`
				)
				.join("")
		);
	}

	render_tabela_atraso(rows) {
		if (!rows || !rows.length) {
			this.$atrasoRows.html(
				`<tr><td colspan="6" class="text-muted">${__("Sem prestações em atraso para os filtros selecionados.")}</td></tr>`
			);
			return;
		}
		this.$atrasoRows.html(
			rows
				.map(
					(r) => `
					<tr>
						<td><a href="/app/pedido-de-credito/${encodeURIComponent(r.parent)}">${frappe.utils.escape_html(r.parent)}</a></td>
						<td>${frappe.utils.escape_html(r.cliente || "")}</td>
						<td>${r.numero}</td>
						<td>${frappe.datetime.str_to_user(r.data_limite_pagamento)}</td>
						<td class="emc-num">${r.dias_atraso}</td>
						<td class="emc-num">${format_currency(r.total_em_atraso, CURRENCY)}</td>
					</tr>`
				)
				.join("")
		);
	}

	inject_styles() {
		if (document.getElementById("emc-painel-style")) return;
		const style = document.createElement("style");
		style.id = "emc-painel-style";
		style.textContent = `
			.emc-painel {
				max-width: 1100px;
				margin: 0 auto;
				padding: 8px 4px 48px;
				font-family: var(--emc-font-numeric);
				color: var(--emc-ink);
			}
			.emc-painel-filtros {
				display: flex;
				flex-wrap: wrap;
				align-items: flex-end;
				gap: 14px;
				background: var(--emc-paper);
				border: 1px solid var(--emc-line);
				border-top: 3px solid var(--emc-accent);
				border-radius: 6px;
				padding: 16px;
				margin-bottom: 18px;
			}
			.emc-painel-filtro {
				min-width: 160px;
			}
			.emc-painel-limpar {
				margin-bottom: 2px;
			}
			.emc-painel-tabelas {
				display: grid;
				grid-template-columns: 1fr 1fr;
				gap: 18px;
				margin-top: 22px;
			}
			.emc-painel-tabelas h4 {
				font-family: var(--emc-font-display);
				margin: 0 0 8px;
				color: var(--emc-accent-deep);
			}
			@media (max-width: 900px) {
				.emc-painel-tabelas {
					grid-template-columns: 1fr;
				}
			}
		`;
		document.head.appendChild(style);
	}
}

File: entre_mc/entre_mc/page/painel_de_credito/painel_de_credito.py
# Copyright (c) 2026, Dércio Bobo and contributors
# For license information, please see license.txt

"""Dados agregados para o Painel de Crédito: cartões (Saldo do Crédito, Dívida,
Em Risco, PAR%, contagens, volume/receita do período) mais duas tabelas de
"precisa de atenção agora" (Carteira em Risco, Créditos em Atraso), todos
filtráveis pela mesma barra de filtros (Cliente/Produto/Finalidade/Datas).

Tudo é recalculado na hora a partir das datas (ver `calcular_saldos`), nunca a
partir do `status` gravado em Plano De Amortizacao - esse só é atualizado pela
tarefa diária `atualizar_atrasos` ou por um Reembolso submetido."""

import frappe
from frappe.utils import date_diff, flt, getdate, nowdate

from entre_mc.entre_mc.doctype.mc_settings.mc_settings import get_settings
from entre_mc.utils.reembolso import calcular_saldos

ESTADOS_CONSIDERADOS = ("Em Curso", "Incumprimento")
LIMITE_LINHAS = 15


@frappe.whitelist()
def obter_painel(cliente=None, produto=None, finalidade=None, data_inicio=None, data_fim=None):
	frappe.has_permission("Pedido De Credito", "read", throw=True)

	settings = get_settings()
	hoje = getdate(nowdate())

	pedido_filters = {"status": ["in", ESTADOS_CONSIDERADOS]}
	if cliente:
		pedido_filters["cliente"] = cliente
	if produto:
		pedido_filters["produto"] = produto
	if finalidade:
		pedido_filters["finalidade"] = finalidade

	pedidos = frappe.get_all(
		"Pedido De Credito", filters=pedido_filters, fields=["name", "cliente", "produto", "status"]
	)

	linhas_por_pedido = {}
	if pedidos:
		linhas = frappe.get_all(
			"Plano De Amortizacao",
			filters={
				"status": ["!=", "Pago"],
				"parenttype": "Pedido De Credito",
				"parent": ["in", [p.name for p in pedidos]],
			},
			fields=[
				"parent",
				"numero",
				"data_limite_pagamento",
				"capital_mensal",
				"capital_pago",
				"juros_mensais",
				"juros_pago",
				"multa_aplicada",
				"multa_paga",
				"juros_mora_aplicado",
				"juros_mora_pago",
			],
		)
		for linha in linhas:
			linhas_por_pedido.setdefault(linha.parent, []).append(linha)

	cards, carteira_em_risco = _cards_e_carteira_em_risco(pedidos, linhas_por_pedido, settings, hoje)
	creditos_em_atraso = _creditos_em_atraso(pedidos, linhas_por_pedido, settings, hoje)
	cards["desembolsado_periodo"], cards["taxas_periodo"] = _totais_do_periodo(
		cliente, produto, finalidade, data_inicio, data_fim
	)

	return {
		"cards": cards,
		"carteira_em_risco": carteira_em_risco[:LIMITE_LINHAS],
		"creditos_em_atraso": creditos_em_atraso[:LIMITE_LINHAS],
	}


def _cards_e_carteira_em_risco(pedidos, linhas_por_pedido, settings, hoje):
	total_saldo = 0
	total_divida = 0
	total_em_risco = 0
	num_ativos = 0
	num_incumprimento = 0
	carteira_em_risco = []

	for pedido in pedidos:
		if pedido.status == "Em Curso":
			num_ativos += 1
		elif pedido.status == "Incumprimento":
			num_incumprimento += 1

		saldo, divida, em_risco = calcular_saldos(linhas_por_pedido.get(pedido.name, []), settings, hoje)
		total_saldo += saldo
		total_divida += divida
		total_em_risco += em_risco

		if divida > 0:
			carteira_em_risco.append(
				{
					"name": pedido.name,
					"cliente": pedido.cliente,
					"saldo_do_credito": saldo,
					"divida": divida,
					"em_risco": em_risco,
				}
			)

	carteira_em_risco.sort(key=lambda d: d["em_risco"], reverse=True)

	cards = {
		"saldo_do_credito": total_saldo,
		"divida": total_divida,
		"em_risco": total_em_risco,
		"par_percent": flt(total_em_risco / total_saldo * 100) if total_saldo else 0,
		"num_ativos": num_ativos,
		"num_incumprimento": num_incumprimento,
	}
	return cards, carteira_em_risco


def _creditos_em_atraso(pedidos, linhas_por_pedido, settings, hoje):
	pedidos_por_nome = {p.name: p for p in pedidos}
	rows = []

	for nome, linhas in linhas_por_pedido.items():
		pedido = pedidos_por_nome.get(nome)
		if not pedido:
			continue
		for linha in linhas:
			dias_atraso = date_diff(hoje, linha.data_limite_pagamento) - flt(settings.dias_de_tolerancia)
			if dias_atraso <= 0:
				continue
			total_em_atraso = flt(
				(linha.capital_mensal - linha.capital_pago)
				+ (linha.juros_mensais - linha.juros_pago)
				+ (linha.multa_aplicada - linha.multa_paga)
				+ (linha.juros_mora_aplicado - linha.juros_mora_pago)
			)
			rows.append(
				{
					"parent": nome,
					"cliente": pedido.cliente,
					"numero": linha.numero,
					"data_limite_pagamento": linha.data_limite_pagamento,
					"dias_atraso": dias_atraso,
					"total_em_atraso": total_em_atraso,
				}
			)

	rows.sort(key=lambda d: d["dias_atraso"], reverse=True)
	return rows


def _totais_do_periodo(cliente, produto, finalidade, data_inicio, data_fim):
	desembolso_filters = {"docstatus": 1}
	if cliente:
		desembolso_filters["cliente"] = cliente
	if data_inicio or data_fim:
		desembolso_filters["data_de_desembolso"] = [
			"between",
			[data_inicio or "1900-01-01", data_fim or "2999-12-31"],
		]

	desembolsos = frappe.get_all(
		"Desembolso",
		filters=desembolso_filters,
		fields=["pedido_de_credito", "valor_desembolsado", "taxa_administrativa"],
	)

	if (produto or finalidade) and desembolsos:
		pedidos_ref = {
			p.name: p
			for p in frappe.get_all(
				"Pedido De Credito",
				filters={"name": ["in", list({d.pedido_de_credito for d in desembolsos})]},
				fields=["name", "produto", "finalidade"],
			)
		}
		if produto:
			desembolsos = [d for d in desembolsos if pedidos_ref.get(d.pedido_de_credito, {}).get("produto") == produto]
		if finalidade:
			desembolsos = [
				d for d in desembolsos if pedidos_ref.get(d.pedido_de_credito, {}).get("finalidade") == finalidade
			]

	total_desembolsado = sum(flt(d.valor_desembolsado) for d in desembolsos)
	total_taxas = sum(flt(d.taxa_administrativa) for d in desembolsos)
	return total_desembolsado, total_taxas
